pkg/liveness/liveness.go

```diff
@@ -0,0 +1,104 @@
+/*
+Copyright 2019 The Ceph-CSI Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package liveness
+
+import (
+	"context"
+	"net"
+	"net/http"
+	"strconv"
+	"time"
+
+	connlib "github.com/kubernetes-csi/csi-lib-utils/connection"
+	"github.com/kubernetes-csi/csi-lib-utils/rpc"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"k8s.io/klog"
+)
+
+var (
+	liveness = prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: "csi",
+		Name:      "liveness",
+		Help:      "Liveness Probe",
+	})
+)
+
+func getLiveness(endpoint string, timeout time.Duration) {
+
+	csiConn, err := connlib.Connect(endpoint)
+	if err != nil {
+		// connlib should retry forever so a returned error should mean
+		// the grpc client is misconfigured rather than an error on the network
+		klog.Fatalf("failed to establish connection to CSI driver: %v", err)
+	}
+
+	klog.Infof("calling CSI driver to discover driver name")
+	csiDriverName, err := rpc.GetDriverName(context.Background(), csiConn)
+	if err != nil {
+		klog.Fatalf("failed to get CSI driver name: %v", err)
+	}
+	klog.Infof("CSI driver name: %q", csiDriverName)
+
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	klog.Infof("Sending probe request to CSI driver %q", csiDriverName)
+	ready, err := rpc.Probe(ctx, csiConn)
+	if err != nil {
+		liveness.Set(0)
+		klog.Errorf("health check failed: %v", err)
+		return
+	}
+
+	if !ready {
+		liveness.Set(0)
+		klog.Error("driver responded but is not ready")
+		return
+	}
+	liveness.Set(1)
+	klog.Infof("Health check succeeded")
+}
+
+func recordLiveness(endpoint string, pollTime, timeout time.Duration) {
+	// register prometheus metrics
+	err := prometheus.Register(liveness)
+	if err != nil {
+		klog.Fatalln(err)
+	}
+
+	// get liveness periodically
+	ticker := time.NewTicker(pollTime)
+	defer ticker.Stop()
+	for range ticker.C {
+		getLiveness(endpoint, timeout)
+	}
+}
+
+func Run(endpoint, livenessendpoint string, port int, pollTime, timeout time.Duration) {
+	klog.Infof("Liveness Running")
+	// start liveness collection
+	go recordLiveness(endpoint, pollTime, timeout)
+
+	// start up prometheus endpoint
+	addr := net.JoinHostPort("0.0.0.0", strconv.Itoa(port))
```

Comment:
instead of `0.0.0.0` can you use below in your deployment template

```
- name: POD_IP
      valueFrom:
        fieldRef:
          fieldPath: status.podIP
```
 to get the pod IP